File: 프로그래머스/2/42860. 조이스틱/조이스틱.js


function solution(name) {
    const n = name.length;
    let answer = Infinity;
    const visited = new Array(n).fill(0);
    const diff = [...name].filter(char => char !== "A").length;
    
    const recur = (depth, now, count, diffCount) => {
        if (diffCount === diff || depth > n) {
            if (diffCount === diff) {
                answer = Math.min(answer, now !== 0 ? count - 1 : count);        
            }
            return;
        }
        const front = name[now].charCodeAt(0) - "A".charCodeAt(0);
        const back = "Z".charCodeAt(0) - name[now].charCodeAt(0) + 1;
        const minDiff = Math.min(front, back);
        
        if (visited[now] === 0) {
            visited[now] = 1;
            recur(depth + 1, 
                  now === n - 1 ? 0 : now + 1, 
                  count + minDiff + 1,
                  minDiff === 0 ? diffCount : diffCount + 1,
                );
            visited[now] = 0;
            
            visited[now] = 1;
            recur(depth + 1, 
                  now === 0 ? n - 1 : now - 1, 
                  count + minDiff + 1,
                  minDiff === 0 ? diffCount : diffCount + 1,
                 );
            visited[now] = 0;
        } else {
            recur(depth + 1, 
                  now === n - 1 ? 0 : now + 1, 
                  count + 1,
                  diffCount
                 );
            recur(depth + 1, 
                  now === 0 ? n - 1 : now - 1, 
                  count + 1,
                  diffCount
                 );
        }
               
    }
    
    recur(0, 0, 0, 0);
    return answer;
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def get_min_diff(char):
    diff1 = ord("Z") - ord(char) + 1
    diff2 = ord(char) - ord("A")
    return min(diff1, diff2)

def solution(name):
    n = len(name)
    vertical = sum(get_min_diff(c) for c in name)
    move = n - 1 

    for i in range(n):
        _next = i + 1
        while _next < n and name[_next] == 'A':
            _next += 1
            
        front = i + i + (n - _next)
        back = 2 * (n - _next) + i
        move = min(move, front, back)

    return vertical + move
